Make client writeFragment and writeQuery behaviour consistent with cache

andrebrantom opened the pull request:
This change makes the API and behaviour of `ApolloClient::writeFragment` consistent with that of `Cache::writeFragment` in that it returns the Reference of the changed object. This is useful if you have some cache manipulation functions that need to run either against the client in a hook or component, or in a mutation `update()` handler.

I created an issue here: https://github.com/apollographql/apollo-feature-requests/issues/319

apollo-cla commented:
@andrebrantom: Thank you for submitting a pull request! Before we can merge it, you'll need to sign the Apollo Contributor License Agreement here: https://contribute.apollographql.com/

benjamn commented on src/core/ApolloClient.ts:
```diff
@@ -439,9 +439,10 @@ export class ApolloClient<TCacheShape> implements DataProxy {
-    this.cache.writeFragment<TData, TVariables>(options);
+  ): Reference | undefined {
+    const ref = this.cache.writeFragment<TData, TVariables>(options);
     this.queryManager.broadcastQueries();
```
```suggestion
    if (options.broadcast !== false) {
      this.queryManager.broadcastQueries();
    }
```
While you're here, here's another alignment/consistency issue. Requesting `broadcast: false` should prevent this `broadcastQueries` call.

andrebrantom replied:
Thanks, I just realised that `writeQuery` suffers the same set of inconsistencies, should update both and re-title the PR?

benjamn replied:
@andrebrantom Yes please!

benjamn reviewed (approved):
LGTM with one more idea for better alignment. Thanks @andrebrantom!

andrebrantom commented:
Apologies for letting this hang, I've just updated both `writeQuery` and `writeFragment` as suggested, I also updated `DataProxy::writeFragment` and `DataProxy::writeQuery` return types to match. Not sure if that might break something, but it means `DataProxy` can now be used to abstract either cache or client for writes.

jerelmiller reviewed (approved):
Thanks a bunch @andrebrantom 🎉 ! We'll get this out with the next patch release!
